sources/web/datalab/polymer/modules/ApiManager.ts

```diff
@@ -289,14 +309,68 @@ class ApiManager {
     const xhrOptions: XhrOptions = {
       method: 'POST',
     }
-    return ApiManager._xhrAsync(ApiManager.terminalApiUrl, xhrOptions);
+    return ApiManager._xhrAsyncBase(ApiManager.terminalApiUrl, xhrOptions);
   }
 
   /**
    * Returns a list of active terminal sessions.
    */
   static listTerminalsAsync() {
-    return ApiManager._xhrAsync(ApiManager.terminalApiUrl);
+    return ApiManager._xhrAsyncBase(ApiManager.terminalApiUrl);
+  }
+
+  static getBasePath() {
+    if (ApiManager.haveBasepath) {
+      return Promise.resolve(ApiManager.basepath)
+    } else {
+      return ApiManager._xhrAsyncRaw(ApiManager.basepathApiUrl).then((response:string) => {
+        const xssiPrefix = ')]}\'\n';
+        if (!response.startsWith(xssiPrefix)) {
+          // Response should be pure text.
+          ApiManager.basepath = response.replace(/\/$/, "");
+          ApiManager.haveBasepath = true;
+          return ApiManager.basepath;
+        } else {
+          response = response.substr(xssiPrefix.length);
+          const j = JSON.parse(response);
+          if (j['basepath']) {
+            ApiManager.basepath = j['basepath'].replace(/\/$/, "");
+            ApiManager.haveBasepath = true;
+            return ApiManager.basepath;
+          } else {
+            ApiManager.xsrfToken = j['token'];
+            var formData = new FormData();
+            formData.append("token", ApiManager.xsrfToken);
+            const xhrOptions: XhrOptions = {
+              noCache: true,
+              method: 'POST',
+              parameters: formData,
+            };
+            return ApiManager._xhrAsyncRaw(ApiManager.basepathApiUrl, xhrOptions).then((response:string) => {
```

Comment:
Convention is to put the `.then` part on the next line, indented by two spaces, makes it a lot more readable.